TG-58 Java new array data in symex

romainbrenguier opened the pull request:
This is to distinguish the statement before and after symex:
`java_new_array` is used before symex, it is then transformed into some
instrtuction containing a `java_new_array_data` by symex.
Symex use to add a new `java_new_array` statement which was confusing.
This PR also improves the coding style in the touched files.

romainbrenguier committed:
Add identifier and rename statement to java_new_array_data

This will be used as an alternative to java_new_array.
This is to distinguish the statement before and after symex:
java_new_array is used before symex, it is then transformed into some
instrtuction containing a java_new_array_data by symex.
Symex use to add a new java_new_array statement which was confusing.

romainbrenguier committed:
Check if object is nil before writing trace

We were accessing fields of the expression without first checking that
it is not nil which could lead to invalid memory access

romainbrenguier committed:
Style: Replace assert by appropriate macros

romainbrenguier committed:
Style: improve documentation in interpreter evaluate

smowton reviewed (approved):
Looks good!

thk123 commented on src/goto-programs/goto_trace.cpp:
```diff
@@ -386,14 +386,8 @@ void show_goto_trace(
       break;
 
     case goto_trace_stept::typet::CONSTRAINT:
-      UNREACHABLE;
```
This kind of change probably at least requires something in the commit message as to why these are now reasonable states to reach (Someone obviously thought they were unreachable before...)

romainbrenguier replied:
`UNREACHABLE` is still present (line 392)

thk123 replied:
Ah of course as it now falls through

thk123 commented on src/goto-programs/goto_trace.cpp:
```diff
@@ -91,7 +91,7 @@ void goto_trace_stept::output(
   out << "\n";
 
-  if(pc->is_other() || pc->is_assign())
+  if((pc->is_other() && lhs_object.is_not_nil())|| pc->is_assign())
```
A good commit message would ideally explain why this condition is now needed. This seems fairly unrelated to this PR and ideally this should have a test demonstrating where this didn't work before.

romainbrenguier replied:
Here is a scenario that was failing before but works now: 
``` c++
SCENARIO("trace.output", "[core]")
{
  symbol_tablet symbol_table;
  namespacet ns(symbol_table);
  goto_programt::instructionst instructions;
  instructions.emplace_back(goto_program_instruction_typet::OTHER);
  goto_trace_stept step;
  step.pc=instructions.begin();
  step.type=goto_trace_stept::typet::ATOMIC_BEGIN;
  step.output(ns, std::cout);
}
```

I could add it as a unit test

thk123 replied:
That seems like a good thing to add if that is considered a reasonable trace :)

thk123 commented on src/goto-programs/builtin_functions.cpp:
```diff
@@ -664,9 +626,8 @@ void goto_convertt::do_java_new_array(
   t_n->code=code_assignt(lhs, malloc_expr);
   t_n->source_location=location;
 
-  assert(ns.follow(object_type).id()==ID_struct);
```
How come this assertion is no longer required.

romainbrenguier replied:
This is already checked by `to_struct_type` on the next line

romainbrenguier replied:
and also line 613

thk123 replied:
I find it is sometimes useful to pull the check out to allow you to document in this specific case why it must be a struct, but this is a preference.

thk123 commented on src/goto-programs/builtin_functions.cpp:
```diff
@@ -664,9 +626,8 @@ void goto_convertt::do_java_new_array(
-  assert(ns.follow(object_type).id()==ID_struct);
   const struct_typet &struct_type=to_struct_type(ns.follow(object_type));
-  assert(struct_type.components().size()==3);
+  PRECONDITION(struct_type.components().size()==3);
```
Consider using `is_valid_java_array` in java types if you want to validate that this is a java Array object (e.g. struct with a  base member, data member and length member).

thk123 commented on src/goto-programs/interpreter_evaluate.cpp:
```diff
@@ -102,7 +102,9 @@ void interpretert::clear_input_flags()
 }
 
-/// \return Number of leaf primitive types; returns true on error
+/// \param ty: a type
```
I'd appreciate a summary on this comment as I don't know what `count_type_leaves` means...

thk123 commented on src/goto-programs/interpreter_evaluate.cpp:
```diff
@@ -299,6 +301,9 @@ bool interpretert::memory_offset_to_byte_offset(
   }
 }
 
+/// Evaluate an expression
```
I'm not sure this summary adds much, consider elaborating on what evaluating an expression means

thk123 reviewed (commented):
There appears to be an unrelated change in this PR that looks like it should at least have a test associated with it.

romainbrenguier committed:
Use existing function for checking object is array

romainbrenguier committed:
Typo in goto_trace output

romainbrenguier committed:
Unit test for goto_trace::output

romainbrenguier committed:
Doc: Summary for count_type_leaves

thk123 commented on unit/goto-programs/goto_trace_output.cpp:
```diff
@@ -0,0 +1,24 @@
+#include <goto-programs/goto_trace.h>
+#include <iostream>
+
+SCENARIO("trace.output", "[core]")
```
Nit: could this be given a better name and be tagged according to the standard:

```
[core][goto-programs][goto_trace]
```

thk123 commented on unit/goto-programs/goto_trace_output.cpp:
```diff
@@ -0,0 +1,4 @@
+//
```
Nit: dunno how this file add ended up in this commit, but it probably doesn't want to be here!

thk123 commented on unit/CMakeLists.txt:
```diff
@@ -26,7 +26,7 @@ list(REMOVE_ITEM sources
 )
 
-add_executable(unit ${sources} ${headers})
+add_executable(unit ${sources} ${headers} goto-programs/goto_trace_output.cpp)
```
I think this does not want to go in as we glob all the cpp files in unit already (@reuk?) (CLion adds this automatically which is possibly something we should see if we can turn off...)

thk123 replied:
I think this might be what is causing the failure on travis...

thk123 reviewed (commented):
This looks good - just a few nits that you may wish to address. Can I request a corresponding test-gen bump to verify it doesn't break anything.

thk123 commented on unit/goto-programs/goto_trace_output.cpp:
```diff
@@ -0,0 +1,25 @@
+/*******************************************************************\
```
Just realised - this still needs to be added to the Makefile as `cbmc` is using Make and CMake in parallel.

romainbrenguier committed:
Add unit test for goto_trace_output in Makefile

thk123 commented:
Looks like a dependency issue in CMake here - but once resolved, assuming passes CI and the TG bump does likewise this looks good to go.

romainbrenguier committed:
Avoid using is_valid_java_array in builin_functions

This is part of another module and we should avoid dependencies.
In particular cmake fails in some cases.

peterschrammel commented on src/util/irep_ids.def:
```diff
@@ -217,6 +217,7 @@ IREP_ID_TWO(cpp_new_array, cpp_new[])
 IREP_ID_TWO(cpp_delete_array, cpp_delete[])
 IREP_ID_ONE(java_new)
 IREP_ID_ONE(java_new_array)
```
Why is java_new_array still needed?

smowton replied:
Previously `ID_java_new_array` was being used to denote **both** the new-array-structure operation (yields something like `{int length; void** data;}*`) **and** the new-plain-array op (which yields a `void**`, like a C++ `new[]`). It was getting away with it because of the order `builtin_functions.cpp` looked for the different cases, but it was very fragile in case that changed. This way the two ops have distinct names (new_array and new_array_data respectively) and it's clear which one is meant.

peterschrammel reviewed (approved):
LGTM
